### ISSUE TITLE
Private divide in active protocol has some instability

### ISSUE THREAD

**Issue by kgoss1729** [cicada-mpc/cicada-mpc#56]
Still thinks that results have been maliciously manipulated (sometimes) when they haven't been. Marked test for this as wip for now till this issue is resolved.

**Comment by kgoss1729** [comment]
Looks like my efforts to correlate randomness between schemes has sorted this out. I just ran a test with 10k divisions and no errors.

**Comment by tshead2** [comment]
If you run ```behave -i active -n "Private Divide -- @2.1"``` repeatedly, it will fail ~50% of the time:

```
Scenario Outline: Private Divide -- @2.1                 # features/active.feature:144
    Given a calculator service with 3 players              # features/steps/calculator_steps.py:58 0.023s
    And a new Active protocol suite                        # features/steps/calculator_steps.py:66 0.014s
    And player 0 secret shares 0.5                         # features/steps/calculator_steps.py:87 0.007s
    And player 1 secret shares 0.3                         # features/steps/calculator_steps.py:87 0.005s
    When the players divide the shares                     # features/steps/calculator_steps.py:223 0.267s
    And the players reveal the secret                      # features/steps/calculator_steps.py:255 0.009s
      Traceback (most recent call last):
        File "/Users/tshead/miniconda3/lib/python3.8/site-packages/behave/model.py", line 1329, in run
          match.run(runner.context)
        File "/Users/tshead/miniconda3/lib/python3.8/site-packages/behave/matchers.py", line 98, in run
          self.func(context, *args, **kwargs)
        File "features/steps/calculator_steps.py", line 257, in step_impl
          _require_success(context.calculator.command("protocol", subcommand="reveal"))
        File "features/steps/calculator_steps.py", line 53, in _require_success
          raise RuntimeError(f"{len(exceptions)} players raised exceptions.")
      RuntimeError: 3 players raised exceptions.
      
      Captured stdout:
      ====== Remote traceback (player 0) ======
      Traceback (most recent call last):
        File "/Users/tshead/src/cicada-mpc/cicada/calculator.py", line 274, in main
          secret = protocol.reveal(share)
        File "/Users/tshead/src/cicada-mpc/cicada/active.py", line 1120, in reveal
          return self._encoder.decode(self._reveal(share, dst=dst))
        File "/Users/tshead/src/cicada-mpc/cicada/active.py", line 1037, in _reveal
          raise ConsistencyError("Secret Shares are inconsistent in the first stage")
      cicada.active.ConsistencyError: Secret Shares are inconsistent in the first stage
      
      ====== Remote traceback (player 1) ======
      Traceback (most recent call last):
        File "/Users/tshead/src/cicada-mpc/cicada/calculator.py", line 274, in main
          secret = protocol.reveal(share)
        File "/Users/tshead/src/cicada-mpc/cicada/active.py", line 1120, in reveal
          return self._encoder.decode(self._reveal(share, dst=dst))
        File "/Users/tshead/src/cicada-mpc/cicada/active.py", line 1037, in _reveal
          raise ConsistencyError("Secret Shares are inconsistent in the first stage")
      cicada.active.ConsistencyError: Secret Shares are inconsistent in the first stage
      
      ====== Remote traceback (player 2) ======
      Traceback (most recent call last):
        File "/Users/tshead/src/cicada-mpc/cicada/calculator.py", line 274, in main
          secret = protocol.reveal(share)
        File "/Users/tshead/src/cicada-mpc/cicada/active.py", line 1120, in reveal
          return self._encoder.decode(self._reveal(share, dst=dst))
        File "/Users/tshead/src/cicada-mpc/cicada/active.py", line 1037, in _reveal
          raise ConsistencyError("Secret Shares are inconsistent in the first stage")
      cicada.active.ConsistencyError: Secret Shares are inconsistent in the first stage
```

**Comment by kgoss1729** [comment]
I'll have to see why that is. I wrote an independent script to test it whilst debugging. There must be some difference between the two that should, hopefully, prove instructive.

**Comment by kgoss1729** [comment]
ב''ה
Greetings,

I have, I think, identified the source of instability in private divide and would apreciate your counsel in how to best address fixing it.

Inside of the untruncated divide function, despite that fact that we call it “untruncated” we do perform some intermediate truncations, we just don’t truncate the final result again. I initially though that there was some instability due to the random bitmasks generated to blind the divisor, but I have confirmed that is not so. I also looked in the size of the errors, which appear to be (relatively speaking) small. Over 100 runs the mean error was -1.35 with standard deviation of 4.68. That sounds like a least significant bit error with some occasional carrying into the next order bits, which I think is in turn a result of the intermediate truncations stocasticity.

The most straigtforward option I see is generating correlated randomness at a higher level of function call so that it is know that the randomness to be used in the lower level calls is the same, then passing that down via optional function arguments which are defaulted to None so that the “normal” use cases are unaffected. Do you have another/better idea or should I proceed on that front? I already do this for part of the division function to make sure that the masks were consistent since that was a huge and obvious problem right away.

שלום וברכה
Peace and Blessing,

Ken Goss, Ph.D.
Sandia National Laboratories
All-Source Analytics Dept., 5552
Location: 750/222
Office: 505-284-3182<tel:1284-3182>
Cell: 417-718-5404
***@***.******@***.***>

From: Timothy M. Shead ***@***.***>
Sent: Friday, August 19, 2022 11:12 AM
To: cicada-mpc/cicada-mpc ***@***.***>
Cc: Goss, Ken ***@***.***>; State change ***@***.***>
Subject: [EXTERNAL] Re: [cicada-mpc/cicada-mpc] Private divide in active protocol has some instability (Issue #56)


If you run behave -i active -n "Private Divide -- @2.1" repeatedly, it will fail ~50% of the time:

Scenario Outline: Private Divide -- @2.1                 # features/active.feature:144

    Given a calculator service with 3 players              # features/steps/calculator_steps.py:58 0.023s

    And a new Active protocol suite                        # features/steps/calculator_steps.py:66 0.014s

    And player 0 secret shares 0.5                         # features/steps/calculator_steps.py:87 0.007s

    And player 1 secret shares 0.3                         # features/steps/calculator_steps.py:87 0.005s

    When the players divide the shares                     # features/steps/calculator_steps.py:223 0.267s

    And the players reveal the secret                      # features/steps/calculator_steps.py:255 0.009s

      Traceback (most recent call last):

        File "/Users/tshead/miniconda3/lib/python3.8/site-packages/behave/model.py", line 1329, in run

          match.run(runner.context)

        File "/Users/tshead/miniconda3/lib/python3.8/site-packages/behave/matchers.py", line 98, in run

          self.func(context, *args, **kwargs)

        File "features/steps/calculator_steps.py", line 257, in step_impl

          _require_success(context.calculator.command("protocol", subcommand="reveal"))

        File "features/steps/calculator_steps.py", line 53, in _require_success

          raise RuntimeError(f"{len(exceptions)} players raised exceptions.")

      RuntimeError: 3 players raised exceptions.



      Captured stdout:

      ====== Remote traceback (player 0) ======

      Traceback (most recent call last):

        File "/Users/tshead/src/cicada-mpc/cicada/calculator.py", line 274, in main

          secret = protocol.reveal(share)

        File "/Users/tshead/src/cicada-mpc/cicada/active.py", line 1120, in reveal

          return self._encoder.decode(self._reveal(share, dst=dst))

        File "/Users/tshead/src/cicada-mpc/cicada/active.py", line 1037, in _reveal

          raise ConsistencyError("Secret Shares are inconsistent in the first stage")

      cicada.active.ConsistencyError: Secret Shares are inconsistent in the first stage



      ====== Remote traceback (player 1) ======

      Traceback (most recent call last):

        File "/Users/tshead/src/cicada-mpc/cicada/calculator.py", line 274, in main

          secret = protocol.reveal(share)

        File "/Users/tshead/src/cicada-mpc/cicada/active.py", line 1120, in reveal

          return self._encoder.decode(self._reveal(share, dst=dst))

        File "/Users/tshead/src/cicada-mpc/cicada/active.py", line 1037, in _reveal

          raise ConsistencyError("Secret Shares are inconsistent in the first stage")

      cicada.active.ConsistencyError: Secret Shares are inconsistent in the first stage



      ====== Remote traceback (player 2) ======

      Traceback (most recent call last):

        File "/Users/tshead/src/cicada-mpc/cicada/calculator.py", line 274, in main

          secret = protocol.reveal(share)

        File "/Users/tshead/src/cicada-mpc/cicada/active.py", line 1120, in reveal

          return self._encoder.decode(self._reveal(share, dst=dst))

        File "/Users/tshead/src/cicada-mpc/cicada/active.py", line 1037, in _reveal

          raise ConsistencyError("Secret Shares are inconsistent in the first stage")

      cicada.active.ConsistencyError: Secret Shares are inconsistent in the first stage

**Comment by kgoss1729** [comment]
fixed better in ddbb8ab0addf58d64a8833390432c3501756cd57
